accepted/2023/feature-switch-attributes.md

```diff
@@ -0,0 +1,718 @@
+
+For example, the feature switch for "dynamic code support" might look like this:
+
+```csharp
+public class RuntimeFeature
+{
+    [FeatureSwitch(typeof(RequiresDynamicCodeAttribute))]
+    public static bool IsDynamicCodeSupported => // ...
+}
+
+
+[FeatureName("System.Runtime.CompilerServices.RuntimeFeature.IsDynamicCodeSupported")]
+class RequiresDynamicCodeAttribute : RequiresFeatureAttribute { }
+```
+
+and a feature guard for "dynamic code compilation" might look like this:
+```csharp
+public class RuntimeFeature
+{
+    [FeatureGuard(typeof(RequiresDynamicCodeAttribute))]
+    public static bool IsDynamicCodeCompiled => // ...
+}
+```
+
+The attribute definitions to support this might look like this:
+```csharp
+[AttributeUsage(AttributeTargets.Class, Inherited = false)]
+abstract class RequiresFeatureAttribute : Attribute { }
+
+[AttributeUsage(AttributeTargets.Property, Inherited = false)]
+class FeatureSwitchAttribute : Attribute {
+    public Type FeatureAttributeType { get; }
+
+    public FeatureSwitchAttribute(Type featureAttributeType) {
+        FeatureAttributeType = featureAttributeType;
+    }
+}
+
+[AttributeUsage(AttributeTargets.Property, Inherited = false, AllowMultiple = true)]
+class FeatureGuardAttribute : Attribute
+{
+    public Type FeatureAttributeType { get; }
+
+    public FeatureGuardAttribute(Type featureAttributeType) {
+        FeatureAttributeType = featureAttributeType;
+    }
+}
+
+[AttributeUsage(AttributeTargets.Class, Inherited = false)]
+class FeatureNameAttribute : Attribute
+{
+    public string Name { get; }
+
+    public FeatureNameAttribute(string name) {
+        Name = name;
+    }
+}
+```
+
+The analyzer would validate that `FeatureGuardAttribute` and `FeatureSwitchAttribute` attribute arguments are derived classes of `FeatureAttribute`, and that `FeatureNameAttribute` is only placed on derived classes of `FeatureAttribute`.
+
+We could use this model even for feature switches similar to `StartupHookSupport`, which don't currently have a `StartupHookSupportAttribute`. There's no need to actually annotate related APIs with the attribute if that is overkill for a small feature. In this case the attribute definition would just serve as metadata that allows us to define a feature switch in a uniform way.
+
+Note that this makes it possible to define custom `Requires` attributes for arbitrary features. While the immediate goal of this proposal is not to enable analyzer support for analysis warnings based on such custom attributes, the model intentionally allows for this so that we could easily open up the analyzer to work for custom features. However, initially support for analysis warnings would be limited to `RequiresUnreferencedCodeAttribute`, `RequiresDynamicCodeAttribute`, and `RequiresAssemblyFilesAttribute`.
```

Comment:
There are actually technical challenges with support custom attributes for custom features - roslyn analyzer has to declare all produced diagnostics, so this would mean that all of the custom attributes would produce the same warning code - which is not ideal. Maybe we'll find a solution to this in the future.